evm/src/cpu/kernel/asm/core/precompiles/expmod.asm

```diff
@@ -1,3 +1,257 @@
+    // stack: 8 * (l_E - 32) + log2(i[96 + l_B..128 + l_B]), retdest
+    SWAP1
+    // stack: retdest, 8 * (l_E - 32) + log2(i[96 + l_B..128 + l_B])
+    JUMP
+case_le_32:
+    // stack: l_E, l_B, retdest
+
+    %log2_floor
+    // stack: log2(l_E), l_B, retdest
+    SWAP2
+    %pop2
+    // stack: log2(l_E), retdest
+    SWAP1
+    // stack: retdest, log2(l_E)
+    JUMP
+
 global precompile_expmod:
-    // TODO
-    PANIC
+    // stack: address, retdest, new_ctx, (old stack)
+    %pop2
+    // stack: new_ctx, (old stack)
+    DUP1
+    SET_CONTEXT
+    // stack: (empty)
+    PUSH 0x100000000 // = 2^32 (is_kernel = true)
+    // stack: kexit_info
+
+    // Load l_B from i[0..32].
+    %stack () -> (@SEGMENT_CALLDATA, 0, 32)
+    // stack: @SEGMENT_CALLDATA, 0, 32, kexit_info
+    GET_CONTEXT
+    // stack: ctx, @SEGMENT_CALLDATA, 0, 32, kexit_info
+    %mload_packing
+    // stack: l_B, kexit_info
+
+    // Load l_E from i[32..64].
+    %stack () -> (@SEGMENT_CALLDATA, 32, 32)
+    GET_CONTEXT
+    %mload_packing
+    // stack: l_E, l_B, kexit_info
+
+    // Load l_M from i[64..96].
+    %stack () -> (@SEGMENT_CALLDATA, 64, 32)
+    GET_CONTEXT
+    %mload_packing
+    // stack: l_M, l_E, l_B, kexit_info
+
+    %stack (l: 3) -> (l, l)
+    // stack: l_M, l_E, l_B, l_M, l_E, l_B, kexit_info
+    %max_3
+    // stack: len, l_M, l_E, l_B, kexit_info
+
+    // Calculate gas costs.
+
+    PUSH l_E_prime_return
+    // stack: l_E_prime_return, len, l_M, l_E, l_B, kexit_info
+    DUP5
+    DUP5
+    // stack: l_E, l_B, l_E_prime_return, len, l_M, l_E, l_B, kexit_info
+    %jump(calculate_l_E_prime)
+l_E_prime_return:
+    // stack: l_E_prime, len, l_M, l_E, l_B, kexit_info
+    DUP5
+    // stack: l_B, l_E_prime, len, l_M, l_E, l_B, kexit_info
+    DUP4
+    // stack: l_M, l_B, l_E_prime, len, l_M, l_E, l_B, kexit_info
+    %max
+    // stack: max(l_M, l_B), l_E_prime, len, l_M, l_E, l_B, kexit_info
+    %expmod_gas_f
+    // stack: f(max(l_M, l_B)), l_E_prime, len, l_M, l_E, l_B, kexit_info
+    SWAP1
+    // stack: l_E_prime, f(max(l_M, l_B)), len, l_M, l_E, l_B, kexit_info
+    PUSH 1
+    %max
+    // stack: max(1, l_E_prime), f(max(l_M, l_B)), len, l_M, l_E, l_B, kexit_info
+    MUL
+    // stack: max(1, l_E_prime) * f(max(l_M, l_B)), len, l_M, l_E, l_B, kexit_info
+    %div_const(3) // G_quaddivisor
+    // stack: (max(1, l_E_prime) * f(max(l_M, l_B))) / G_quaddivisor, len, l_M, l_E, l_B, kexit_info
+    PUSH 200
+    %max
+    // stack: g_r, len, l_M, l_E, l_B, kexit_info
+    %charge_gas
+
+    // stack: len, l_M, l_E, l_B, kexit_info
+
+    // Copy B to kernel general memory.
+    DUP4
+    // stack: l_B, len, l_M, l_E, l_B, kexit_info
+    PUSH 96
+    PUSH @SEGMENT_CALLDATA
+    GET_CONTEXT
+    PUSH 0
+    PUSH @SEGMENT_KERNEL_GENERAL
+    PUSH 0
+    // stack: dst=(0, @SEGMENT_KERNEL_GENERAL, b_loc=0), src=(ctx, @SEGMENT_CALLDATA, 96), l_B, len, l_M, l_E, l_B, kexit_info
+    %memcpy
+    // stack: len, l_M, l_E, l_B, kexit_info
```

Comment:
Looking at the debug trace, this stack comment isn't correct. In particular, `kexit_info` is in front of the stack at this point.

Reply:
Might be fixed... but what test do I run to see that?

Reply:
Yes it's fixed 👍 
I'm running the `stReturnDataTest/modexp_modsize0_returndatasize_0` test.